### PR TITLE
feat (ai/core): add experimental tool call repair

### PR DESCRIPTION
**Pull request by lgrammel**
## Tasks

- [x] implementation
   - [x] expose ToolCallRepairFunction
   - [x] parse tool call
      - [x] implementation
      - [x] test
   - [x] generateText
   - [x] streamText
   - [x] tool call repair error
- [x] examples
- [x] documentation
   - [x] tool calling guide error handling section
   - [x] tool calling guide repair section
   - [x] generateText reference
   - [x] streamText reference
   - [x] tool call repair error
- [x] changeset

### CODE REVIEW

**Comment by williamlmao** [comment]
👀 

This looks interesting! I would honestly still rather just have the ability to catch the full response from the LLM provider and have the flexibility to deal with things however I want. Abstractions and helpers like this are awesome, but if we can _only_ access the provider response through these then the SDK starts to feel rigid / over abstracted. I think having both is the best case though - an easy helper for those who want to use it, but ultimately flexibility for someone who needs to do something more custom.

I would love to know where your head is regarding my last comment on #3779, I am happy to put in a PR to expose the full provider response on the error object if you are aligned. Thanks!

**Comment by lgrammel** [comment]
@williamlmao the approach that you propose has several limitations / downsides:

a) full responses are not always available (streaming, or providers w/o api calls)
b) for correct multi-step behavior, the error recovery / tool call repair needs to be handled inside generateText/streamText
c) exposing provider-specific results needs to dependencies that the AI SDK is trying to abstract, namely code that relies on provider-specific details

I want to explore tool call repair first (and if it works structured output repair as well), before looking into a more provider specific approach if needed. My hunch is that tool call repair & structured output repair will end up being sufficient and better integrated.

File: content/docs/03-ai-sdk-core/15-tools-and-tool-calling.mdx [hunk 1]
```diff
@@ -396,6 +397,67 @@ return result.toDataStreamResponse({
...
+  The tool call repair feature is experimental and may change in the future.
+</Note>
+
+Language models sometimes fail to generate valid tool calls,
+especially when the parameters are complex or the model is smaller.
+
+You can use the `experimental_toToolCallRepair` function to attempt to repair the tool call
+with a custom function.
+
+You can use different strategies to repair the tool call, e.g.:
```
Comment (nicoalbanese) [inline content/docs/03-ai-sdk-core/15-tools-and-tool-calling.mdx:412]:
```suggestion
You can use different strategies to repair the tool call:
```

File: content/docs/07-reference/05-ai-sdk-errors/ai-tool-call-repair-error.mdx [hunk 2]
```diff
@@ -0,0 +1,28 @@
+---
+title: ToolCallRepairError
+description: Learn how to fix ToolCallRepairError
```
Comment (nicoalbanese) [inline content/docs/07-reference/05-ai-sdk-errors/ai-tool-call-repair-error.mdx:3]:
```suggestion
description: Learn how to fix AI SDK ToolCallRepairError
```

**Review by nicoalbanese — APPROVED** [review]

**Pull request merged**